### DISCUSSION TITLE
Single-trial dynamics, bstep error and differences between models B/C/D

### DISCUSSION THREAD

**Discussion by JulieB1234** [cvnlab/GLMsingle#160] [Q&A]
![figure1](https://github.com/user-attachments/assets/8a2a039e-03e1-4c1d-9da9-139fe9753ddd)
<img width="1362" alt="figure2" src="https://github.com/user-attachments/assets/35373685-5e09-4305-942e-8357cae7a837">
<img width="1366" alt="figure3" src="https://github.com/user-attachments/assets/cfc58a22-c1ee-4eb5-bd88-cef41e3e62af">

I am working on fMRI data on the topic of auditory perception around perception threshold in healthy volunteers. My experiment’s setting is an event-related one where participants are presented with an auditory stimulus which varies in intensity across trials (5 different levels, named « SNR1 » to « SNR5 », the « SNR1 »  being actually no sound at all, and SNR3-SNR4 being intermediate intensities around the perceptual threshold for which the stimulus is sometimes perceived sometimes not).
I would like to investigate the single-trial dynamics of brain activation elicited by these stimuli. To do so I have used the GLMsingle toolbox on my design matrix (using Matlab and SPM12-based design matrices) and, for now at individual level only, I have plotted the single trial beta estimates as a function of stimulus intensity (ie, SNR-level) in pre defined regions of interest (I join an example -figure 1- of such figures where we plot: (1) scatter plot of the single-trial beta values for each SNR ; (2) probability density / distribution of single trial beta values for each SNR ; (3) mean of single-trial beta values for each SNR ; and (4) inter-trial variability measured by standard deviation of single trial beta values for each SNR).

I have a few questions regarding the results I got :

1/ during the C and D models (GLMdenoise and fracridge regression) I get an error during alpha grid specification in the fracridge function, telling me to decrease bstep because « mxgap<0.2 » but actually decreasing step didn’t solve it and I found out the problem of mxgap being too big is only the case for a few ii values that correspond to voxels outside of the brain ; furthermore, for some of these voxels (but not all), continuing into the GLMsetimatesingletrial.m function (after fracridge function ; line 1423, during olsmatrix function) I get a warning: "Matrix is singular, close to singular or badly scaled. Results may be inaccurate. RCOND = NaN. »
What do you think of these errors? Are they just outlier values outside of the brain that we should not take into account? How do you explain this doesn’t happen for all the « outside brain » voxels ?

2/ depending on the model used (B, B+C, or B+C+D) the results change pretty strongly, as you can see in figures 2 and 3 compared to figure 1 (all are for the same subject and region of interest, only the model changes) ; how would you explain it and how to choose the best model(s) for our data?
NB: we already have many nuisance regressors based on physiological recordings during scanning so maybe it is redundant with GLMdenoise?

If you could help me understand these points it would be very helpful, 
Thank you for your time

**Comment by kendrickkay** [comment]
Hello - sorry for the delay.

A few thoughts:
- To understand better, it would be helpful to see the design matrix you are giving to GLMsingle, as well as any options/flags that you are using.
- Regarding the bstep issue, I am not sure which line of code you are referring to (line 1423 does not correspond to my version).  Can you show the error message and/or provide some way to reproduce the error?  
- Voxels clearly outside the brain can be ignored, for sure.  But, I still want to know why the code is failing (it should not fail, in theory).  Can you see what the data are like for those voxels?  Are they all 0 in the time series?
- Regarding the results changing across the versions, it is good that you are carefully inspecting the results. However, based on what I see, the amount of change isn't that much (like it's not crazy). Furthermore, the whole point of GLMsingle is to produce better quality beta estimates...
- You mention nuisance regressors --- can I ask how exactly you are using them / including them in your data processing pipeline?
- If you could provide the full workspace (data and design matrix, the call the GLMsingle) that can help diagnose what is going on.

Kendrick

**Comment by JulieB1234** [comment]
Hello

Thank you for your reply.
I am very sorry for the delay of response; in the end we solved the problem and the toolbox works well for us

I have one question however: are the single trial beta values z-scored at some point ? (If using all models together) it indeed seems like they are when looking at the final data, but I haven’t found an obvious part of script where it happens

Thank you 
Best regards, 

Julie

**Comment by kendrickkay** [comment]
Great.  GLMsingle does not itself perform z-scoring of the final single trial beta weights.  The final weights are provided in units of percent BOLD signal change.  However, we suggest that users may want to perform z-scoring of the single trial beta weights for each voxel across each scan session. The primary motivation would be to reduce nonstationarities across different scan sessions.  This is discussed to some degree in the NSD paper.  So, if you are combining data across multiple scan sessions, this may be something to consider.   

When you say "seems like they are when looking at the final data", what in particular are you referring to?

**Comment by JulieB1234** [comment]
Thanks for your answer
I have been studying single-trial betas distribution for different conditions (different levels of stim intensity from snr1 to 5) and comparing the GLMsingle model (B+C+D) with a simple single-trial design matrix run in SPM so a classic OLS

And the beta values differ, cf. figures to illustrate it in a given ROI: 1st one is with GLMsingle (plots purple), 2nd one with spm (plots dark blue and thicker) for both ROI shown here
Also, how would you interpret the negative value for the second ROI ?
￼￼
￼
￼

**Comment by kendrickkay** [comment]
The images didn't come through; can you edit/modify?

**Comment by JulieB1234** [comment]
![SingleTrialGLMsubj2Active_ROIcustom5mm_10voxels_leftSTG_A22c](https://github.com/user-attachments/assets/136dbdf6-4272-4638-b464-e3c9d0510f5d)
![SingleTrialSPMsubj2Active_ROIcustom5mm_10voxels_leftSTG_A22c](https://github.com/user-attachments/assets/27e534bf-dc97-4101-8ea8-e12aaba7ec5d)

![SingleTrialGLMsubj2Active_ROIcustom5mm_30voxels_leftIns_dIa](https://github.com/user-attachments/assets/8d849c3c-1d36-40e8-bc26-d944db7fd250)
![SingleTrialSPMsubj2Active_ROIcustom5mm_30voxels_leftIns_dIa](https://github.com/user-attachments/assets/5e731fa8-8e2b-4400-bad9-2741ef20a6a3)

**Reply by kendrickkay** [reply]
A few comments:
- Regarding the thick blue line results (SPM), I am not sure about the units of the betas that you are getting.  In the first ROI, the values of around 6-10 are quite high --- are you trying to interpret those numbers as the amplitude of the BOLD response in units of percent signal change?   That is what GLMsingle returns, but if it's not exactly what you are getting with SPM, then they won't be directly comparable.
- Regarding the negative betas in the second ROI, it could be real.  It is known that brain regions can show a decrease in the BOLD response when an experimental manipulation occurs.  I guess it depends on what exactly your experiment is.   Also, note that increases/decreases also depend on accurately estimating the baseline signal level, and therefore also depends on what is taken to be the baseline in your experiment.   If there are difficulties or inaccuracies in estimating your baseline signal, that will influence the magnitudes/signs of your betas.

**Comment by JulieB1234** [comment]
Okay I’ll look into it
Thanks a lot for your help Kendrick!
